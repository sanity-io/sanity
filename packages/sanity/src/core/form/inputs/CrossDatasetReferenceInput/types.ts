import {PreviewValue} from '@sanity/types'
import {Observable} from 'rxjs'
import {DocumentAvailability} from '../../../preview'

/** @internal */
export interface CrossDatasetReferenceInfo {
  id: string
  type: string | undefined
  availability: DocumentAvailability | null
  preview: {
    published: PreviewValue | undefined
  }
}

/** @internal */
export interface SearchState {
  hits: CrossDatasetSearchHit[]
  searchString?: string
  isLoading: boolean
}

/** @internal */
export type CrossDatasetSearchFunction = (query: string) => Observable<CrossDatasetSearchHit[]>

/** @internal */
export interface CrossDatasetSearchHit {
  id: string
  type: string
  published: undefined | {_id: string; _type: string}
}
