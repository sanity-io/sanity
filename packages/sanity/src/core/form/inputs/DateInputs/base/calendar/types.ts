export interface CalendarLabels {
  goToTomorrow: string
  goToToday: string
  goToYesterday: string
  goToPreviousYear: string
  goToNextYear: string
  goToPreviousMonth: string
  goToNextMonth: string
  selectHour: string
  selectMinute: string
  setToCurrentTime: string
  monthNames: MonthNames
  weekDayNamesShort: WeekDayNames
  setToTimePreset: (time: string, date: Date) => string
}

export type WeekDayNames = [
  mon: string,
  tue: string,
  wed: string,
  thu: string,
  fri: string,
  sat: string,
  sun: string,
]

export type MonthNames = [
  jan: string,
  feb: string,
  mar: string,
  apr: string,
  may: string,
  jun: string,
  jul: string,
  aug: string,
  sep: string,
  oct: string,
  nov: string,
  dec: string,
]
