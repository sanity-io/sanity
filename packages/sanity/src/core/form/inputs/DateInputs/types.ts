export type ParseResult = {isValid: boolean; date?: Date; error?: string} & (
  | {isValid: true; date: Date}
  | {isValid: false; error?: string}
)
