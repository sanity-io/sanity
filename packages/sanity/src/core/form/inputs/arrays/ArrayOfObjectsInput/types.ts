/* eslint-disable camelcase */

import type {ComponentType} from 'react'
import {KeyedSegment} from '@sanity/types'

export type ModalType = 'modal' | 'fullscreen' | string

/** @public */
export type ArrayInputMember = {
  _type?: string
  _key: string
  [key: string]: any
}

/** @public */
export interface InsertEvent {
  item: ArrayInputMember
  position: 'before' | 'after'
  referenceItem: KeyedSegment | number
  edit?: boolean
}

/** @internal */
export type ReferenceItemComponentType = ComponentType<any>
