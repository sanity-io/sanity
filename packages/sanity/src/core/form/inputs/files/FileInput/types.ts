import {type AssetFromSource, type AssetSource} from '@sanity/types'

import {type FileInfo} from '../common/styles'
import {type BaseFileInputProps} from './FileInput'

export interface FileAssetProps extends Omit<BaseFileInputProps, 'renderDefault'> {
  browseButtonElementRef: React.RefObject<HTMLButtonElement | null>
  clearField: () => void
  hoveringFiles: FileInfo[]
  isBrowseMenuOpen: boolean
  isStale: boolean
  isUploading: boolean
  onCancelUpload?: () => void
  onClearUploadStatus: () => void
  onSelectAssets: (assetsFromSource: AssetFromSource[]) => void
  onSelectFiles: (assetSource: AssetSource, files: File[]) => void
  onStale: () => void
  selectedAssetSource: AssetSource | null
  setBrowseButtonElement: (element: HTMLButtonElement | null) => void
  setHoveringFiles: (hoveringFiles: FileInfo[]) => void
  setIsBrowseMenuOpen: (isBrowseMenuOpen: boolean) => void
  setIsUploading: (isUploading: boolean) => void
  setSelectedAssetSource: (assetSource: AssetSource | null) => void
}
