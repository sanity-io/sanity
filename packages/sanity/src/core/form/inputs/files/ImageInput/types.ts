import {type SanityClient} from '@sanity/client'
import {
  type AssetSource,
  type Image as BaseImage,
  type ImageAsset,
  type ImageSchemaType,
  type UploadState,
} from '@sanity/types'
import {type Observable} from 'rxjs'

import {type UploaderResolver} from '../../../studio/uploads/types'
import {type ObjectInputProps} from '../../../types'
import {type ImageUrlBuilder} from '../types'

/**
 * @hidden
 * @internal
 */
export type FileInfo = {
  type: string // mime type
  kind: string // 'file' or 'string'
}

/**
 * @hidden
 * @beta
 */
export interface BaseImageInputValue extends Partial<BaseImage> {
  _upload?: UploadState
}

/**
 * @hidden
 * @beta */
export interface BaseImageInputProps
  extends ObjectInputProps<BaseImageInputValue, ImageSchemaType> {
  assetSources: AssetSource[]
  directUploads?: boolean
  imageUrlBuilder: ImageUrlBuilder
  observeAsset: (documentId: string) => Observable<ImageAsset>
  resolveUploader: UploaderResolver
  client: SanityClient
  t: (key: string, values?: Record<string, string>) => string
}
