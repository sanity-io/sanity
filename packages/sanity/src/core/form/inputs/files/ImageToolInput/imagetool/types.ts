import type {Rect} from './2d/shapes'

export type CropMethod = 'full_width' | 'letterbox' | 'full_height'

export interface Hotspot {
  x: number
  y: number
  height: number
  width: number
}

export interface Crop {
  top: number
  right: number
  bottom: number
  left: number
}

export interface EdgeOffsets {
  top: number
  right: number
  bottom: number
  left: number
}

export interface CropAndHotspot {
  hotspot: Hotspot
  crop: Crop
}

export interface Offsets {
  top: number
  right: number
  bottom: number
  left: number
}

export interface Dimensions {
  width: number
  height: number
}

export interface Coordinate {
  x: number
  y: number
}

export interface CropHandles {
  left: Rect
  right: Rect
  top: Rect
  topLeft: Rect
  topRight: Rect
  bottom: Rect
  bottomLeft: Rect
  bottomRight: Rect
}

export interface ToolCanvasProps {
  value: Partial<CropAndHotspot>
  image: HTMLCanvasElement
  onChange: (value: {hotspot: Hotspot} | {crop: Crop}) => void
  onChangeEnd: (value: {hotspot: Hotspot} | {crop: Crop}) => void
  readOnly: boolean
}

export interface ToolCanvasState {
  cropping: keyof CropHandles | false
  resizing: boolean
  moving: boolean
  cropMoving: boolean
  mousePosition: Coordinate | null
}
