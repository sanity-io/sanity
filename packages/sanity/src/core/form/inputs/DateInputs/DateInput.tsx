import React, {useCallback} from 'react'
import {format, parse} from '@sanity/util/legacyDateFormat'
import {set, unset} from '../../patch'
import {StringInputProps} from '../../types'
import {CommonDateTimeInput} from './CommonDateTimeInput'

interface ParsedOptions {
  dateFormat: string
  calendarTodayLabel: string
}

interface SchemaOptions {
  dateFormat?: string
  calendarTodayLabel?: string
}

/**
 * @hidden
 * @beta */
export type DateInputProps = StringInputProps

// This is the format dates are stored on
const VALUE_FORMAT = 'YYYY-MM-DD'
// default to how they are stored
const DEFAULT_DATE_FORMAT = VALUE_FORMAT

function parseOptions(options: SchemaOptions = {}): ParsedOptions {
  return {
    dateFormat: options.dateFormat || DEFAULT_DATE_FORMAT,
    calendarTodayLabel: options.calendarTodayLabel || 'Today',
  }
}

const deserialize = (value: string) => parse(value, VALUE_FORMAT)
const serialize = (date: Date) => format(date, VALUE_FORMAT)

/**
 * @hidden
 * @beta */
export function DateInput(props: DateInputProps) {
  const {readOnly, onChange, schemaType, elementProps, value} = props
  const {dateFormat} = parseOptions(schemaType.options)

  const handleChange = useCallback(
    (nextDate: string | null) => {
      onChange(nextDate === null ? unset() : set(nextDate))
    },
    [onChange],
  )

  const formatInputValue = useCallback((date: Date) => format(date, dateFormat), [dateFormat])

  const parseInputValue = useCallback(
    (inputValue: string) => parse(inputValue, dateFormat),
    [dateFormat],
  )

  return (
    <CommonDateTimeInput
      {...elementProps}
      deserialize={deserialize}
      formatInputValue={formatInputValue}
      onChange={handleChange}
      parseInputValue={parseInputValue}
      placeholder={schemaType.placeholder}
      readOnly={readOnly}
      selectTime={false}
      serialize={serialize}
      value={value}
    />
  )
}
