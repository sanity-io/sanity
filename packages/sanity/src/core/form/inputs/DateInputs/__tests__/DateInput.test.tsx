import {defineField} from '@sanity/types'
import {fireEvent} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import {renderStringInput} from '../../../../../../test/form'
import {DateInput} from '../DateInput'

// NOTE: for the tests to be deterministic we need this to ensure tests are run in a predefined timezone
// see globalSetup in jest config for details about how this is set up
test('timezone for the test environment should be set to America/Los_Angeles', () => {
  expect(Intl.DateTimeFormat().resolvedOptions().timeZone).toBe('America/Los_Angeles')
})

test('does not emit onChange after invalid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'date',
      name: 'test',
    }),
    render: (inputProps) => <DateInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  userEvent.type(input, 'this is invalid')
  expect(input.value).toBe('this is invalid')
  expect(onChange.mock.calls.length).toBe(0)

  fireEvent.blur(input)

  expect(onChange.mock.calls.length).toBe(0)
})

test('emits onChange on correct format if a valid value has been typed', async () => {
  const {onChange, result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'date',
      name: 'test',
    }),
    render: (inputProps) => <DateInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // NOTE: the date is entered and displayed in local timezone
  userEvent.type(input, '2021-03-28')
  expect(input.value).toBe('2021-03-28')

  fireEvent.blur(input)

  // NOTE: the date is entered and displayed in local timezone but stored in utc
  expect(onChange.mock.calls).toMatchSnapshot()
})

test('formatting of deserialized value', async () => {
  const {result} = await renderStringInput({
    fieldDefinition: defineField({
      type: 'date',
      name: 'test',
    }),
    props: {documentValue: {test: '2021-03-28'}},
    render: (inputProps) => <DateInput {...inputProps} />,
  })

  const input = result.container.querySelector('input')!

  // const {textInput} = renderInput({value: '2021-03-28'} as any)
  expect(input.value).toBe('2021-03-28')
})
