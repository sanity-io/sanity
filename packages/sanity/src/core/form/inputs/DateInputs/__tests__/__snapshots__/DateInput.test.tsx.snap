// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emits onChange on correct format if a valid value has been typed 1`] = `
Array [
  Array [
    Object {
      "patchType": Symbol(sanity.patch),
      "path": Array [],
      "type": "set",
      "value": "2021-03-28",
    },
  ],
]
`;
